lara-typescript/src/interactive-api-parent/modal-api-plugin.ts

```diff
@@ -0,0 +1,145 @@
+          alert: (opts: ICloseModal) => closeAlert(opts),
+          lightbox: (opts: ICloseModal) => closeLightbox(opts),
+          dialog: (opts: ICloseModal) => closeDialog(opts)
+        };
+        const type = modalMap[options.uuid]?.type;
+        const closeFn = type && fnMap[type];
+        closeFn?.(options);
+      }
+    }
+  };
+}
+
+function showAlert(options: IShowAlert) {
+  const { style, title: _title, text } = options;
+  let title: string;
+  let titlebarColor: string | undefined;
+  let message: string;
+  if (style === "correct") {
+    title = _title != null ? _title : "Correct";
+    titlebarColor = "#75a643";
+    message = text || "Yes! You are correct.";
+  }
+  else if (style === "incorrect") {
+    title = _title != null ? _title : "Incorrect";
+    titlebarColor = "#b45532";
+    message = text || "Sorry, that is incorrect.";
+  }
+  else {
+    title = _title || "";
+    message = text || "";
+  }
+
+  const $content = $(`<div class='check-answer'><p class='response'>${message}</p></div`);
+  modalMap[options.uuid] = { type: options.type, $content };
+  addPopup({
+    content: $content[0],
+    title,
+    titlebarColor,
+    modal: true,
+    onClose: () => delete modalMap[options.uuid]
+  });
+}
+
+function closeAlert(options: ICloseModal) {
+  const $content = modalMap[options.uuid]?.$content;
+  if ($content?.is(":ui-dialog")) {
+    $content.dialog("close");
+  }
+}
+
+function showLightbox(options: IShowLightbox) {
+  const getSizeOptions = () => {
+    const { isImage, size } = options;
+    // colorbox implementation detail
+    const kChromeSize = { width: 42, height: 70 };
+    const availableWidth = window.innerWidth - kChromeSize.width;
+    const availableHeight = window.innerHeight - kChromeSize.height;
+    const sizeOpts: { width?: number, height?: number, scalePhotos?: boolean } = {};
+    // images are auto-sized correctly without intervention
+    if (isImage) {
+      // prevent scaling up unless requested
+      const requiresUpscaling = (!!size?.width && size.width < availableWidth) &&
+                                (!!size?.height && size.height < availableHeight);
+      sizeOpts.scalePhotos = !requiresUpscaling || options.allowUpscale;
+      return sizeOpts;
+    }
+    // for iframes, if size is specified, then maintain aspect ratio
+    if (size) {
+      const aspectRatio = size.width / size.height;
+      if (size.height > availableHeight) {
+        // height is constraining dimension
+        sizeOpts.width = availableHeight * aspectRatio;
+        sizeOpts.height = availableHeight;
+      }
+      else {
+        sizeOpts.width = size.width;
+        sizeOpts.height = size.height;
+      }
+      if (size.width > availableWidth) {
+        // width is constraining dimension
+        sizeOpts.width = availableWidth;
+        sizeOpts.height = availableWidth / aspectRatio;
+      }
+      return sizeOpts;
+    }
```

Comment:
It might not be necessary at first, but in other cases we try to handle users who resize their browser. So that means this function will need to re-called on resize events.

Also interactive iframes have the ability to request a height or aspect ratio depending on what a user does. In this case it doesn't look like we are connecting iframephone to the iframe in the lightbox.  

However I think that also means that an image interactive that is showing its own magnifying glass to close it self, will have have no way to tell LARA to close itself. So it seems like we will need the iframphone setup.

Reply:
Resizability seems like a new feature/story. It's not a capability of the existing image question lightbox, for instance. It seems as though we ought to be able to handle browser window resize automatically without additional API.

The existing image lightbox implementation already handles click on the image to close the lightbox, so that functionality is already present. Making it possible for an `<iframe>`d interactive to close the lightbox seems perfectly reasonable but my inclination would be to make it a separate story as well.

Reply:
Yeah I think it is fine to leave it this way, and we'll see how the dialog stuff works out. My original intention was not to support images just iframes. Which means they'd need to a way to close the lightbox.

The current resizing of iframes doesn't require an additional API for browser size changes. The interactive just provides a height or aspectRatio, and the LARA side of things does the rest of the computations. But at the same since the user can change the content inside of the iframe, and we normally want to avoid putting the scroll bars inside of the iframe itself, so we have an API for the interactive to request more height from LARA.

In any case you will run into all of this when implementing the dialog. Hopefully the interactive sizing will be a little bit cleaner then so it will be easier to understand from the code.